### PR TITLE
rocAL - adding generic plugin for GPU support 

### PR DESCRIPTION
**Pull request by LakshmiKumar23**
This PR allows to run rocAL on GPU without torch installed.
Modified the python unit test to use this plugin.
Extra argument added to unit test to take in ```--device-id``` for systems with multiple GPUs

### CODE REVIEW

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 1]
```diff
@@ -112,7 +113,7 @@ def __init__(self, batch_size=-1, num_threads=-1, device_id=-1, seed=-1,
         self._check_ops = ["CropMirrorNormalize"]
         self._check_crop_ops = ["Resize"]
         self._check_ops_decoder = ["ImageDecoder", "ImageDecoderSlice" , "ImageDecoderRandomCrop", "ImageDecoderRaw"]
-        self._check_ops_reader = ["FileReader", "TFRecordReaderClassification", "TFRecordReaderDetection",
+        self._check_ops_reader = ["labelReader", "TFRecordReaderClassification", "TFRecordReaderDetection",
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py:116]: why is this changed? FileReader seems correct

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py [hunk 2]
```diff
@@ -267,7 +282,12 @@ def GetBBCords(self, array):
         return b.getBBCords(self._handle, array)
 
     def getImageLabels(self, array):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/pipeline.py:284]: Do we still need GetImageLables. If not please remove that from function definitions

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 1 above)
Comment (LakshmiKumar23) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py:116]: I thought so too. [Here](https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/blob/master/rocAL/rocAL_pybind/amd/rocal/readers.py#L45) it takes it as labelReader, and I do not see a FileReader

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 2 above)
Comment (LakshmiKumar23) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py:284]: I wanted to wait for MCW to respond before removing it

File: rocAL/rocAL_pybind/amd/rocal/plugin/generic.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            if self.device == "cpu":
+                if self.tensor_dtype == types.FLOAT:
+                    self.out = np.empty((self.bs*self.n, self.p, int(self.h/self.bs), self.w,), dtype=np.float32)
+                elif self.tensor_dtype == types.FLOAT16:
+                    self.out = np.empty((self.bs*self.n, self.p, int(self.h/self.bs), self.w,), dtype=np.float16)
+                self.labels = np.empty(self.labels_size, dtype = np.int32)
+
+            else:
+                gpu_device = cp.cuda.Device(self.device_id)
+                gpu_device.use()
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/plugin/generic.py:90]: probably better to use "with Device(id):" instead of use()

File: rocAL/rocAL_pybind/amd/rocal/plugin/generic.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            if self.device == "cpu":
+                if self.tensor_dtype == types.FLOAT:
+                    self.out = np.empty((self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype=np.float32)
+                elif self.tensor_dtype == types.FLOAT16:
+                    self.out = np.empty((self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype=np.float16)
+                self.labels = np.empty(self.labels_size, dtype = np.int32)
+
+            else:
+                gpu_device = cp.cuda.Device(self.device_id)
+                gpu_device.use()
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/plugin/generic.py:107]: same as prev comment

File: rocAL/rocAL_pybind/examples/parse_config.py [hunk 5]
```diff
@@ -63,6 +63,8 @@ def parse_args():
     common_group.add_argument('--fp16', default=False, action='store_true',
                         help='run input pipeline fp16 format')
     
+    common_group.add_argument('--device-id', type=int, default=0,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/examples/parse_config.py:66]: I think local_rank is used to identify the device_id

File: rocAL/rocAL_pybind/examples/parse_config.py [hunk 6]
```diff
@@ -63,6 +63,8 @@ def parse_args():
...
+                        help='choose the GPU device for rocal GPU mode')
     common_group.add_argument('--local-rank', type=int, default=0,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/examples/parse_config.py:68]: explanation for local_rank, world_size and num_threads is not correct

File: rocAL/rocAL_pybind/rocal_pybind.cpp [hunk 7]
```diff
@@ -147,6 +147,36 @@ namespace rocal{
         return py::cast<py::none>(Py_None);
     }
 
+    py::object wrapper_cupy_tensor32(RocalContext context, size_t array_ptr,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/rocal_pybind.cpp:150]: can you please rename to "wrapper_copy_cupy_tensor32"

File: rocAL/rocAL_pybind/rocal_pybind.cpp [hunk 8]
```diff
@@ -147,6 +147,36 @@ namespace rocal{
...
+        float * ptr = (float*)array_ptr;
+        // call pure C++ function
+        int status = rocalCopyToOutputTensor32(context, ptr, tensor_format, multiplier0,
+                                              multiplier1, multiplier2, offset0,
+                                              offset1, offset2, reverse_channels);
+        // std::cerr<<"\n Copy failed with status :: "<<status;
+        return py::cast<py::none>(Py_None);
+    }
+
+    py::object wrapper_cupy_tensor16(RocalContext context, size_t array_ptr,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/rocal_pybind.cpp:165]: change the name as suggested before

**Review by rrawther — CHANGES_REQUESTED** [review]
@LakshmiKumar23 : Since we change the unit_test to use generic api: we need to add another test to test pytorch iterator

File: rocAL/rocAL_pybind/amd/rocal/pipeline.py (hunk 1 above)
Comment (LakshmiKumar23) [reply rocAL/rocAL_pybind/amd/rocal/pipeline.py:116]: Confirmed with @swetha097, it's the name that's there

**Comment by LakshmiKumar23** [comment]
> @LakshmiKumar23 : Since we change the unit_test to use generic api: we need to add another test to test pytorch iterator

@rrawther Added the old file back for pytorch iterator and renamed file

**Review by rrawther — APPROVED** [review]
